### DISCUSSION TITLE
Frizbee fuzzy matcher

### DISCUSSION THREAD

**Discussion by saghen** [alexpasmantier/television#19] [Ideas]
Hey there, love the project! I recently wrote a fuzzy matching library in SIMD that outperforms nucleo and includes many of the scoring bonuses from FZY. It also optionally allows for typos by disabling the prefiltering step in nucleo/fzf. If you're interested in testing it, I'd be more than happy to throw up a PR integrating it!

https://github.com/Saghen/frizbee

**Comment by alexpasmantier** [comment]
Hi there, thanks for the interest!

Nice work on that SIMD matcher.
Sounds like an exciting experiment to try and swap out nucleo for frizbee indeed.

Just ran a couple of benches locally and it indeed seems consistently faster by a factor >2. 

The only concern I have is that the high-level [`nucleo`](https://docs.rs/nucleo/latest/nucleo/) crate also provides quite a fair amount of plumbing around the matcher that enables things like incremental inspection of the matcher state while it's running (which means television can for example feed results to the UI as they arrive), optimizations when its current pattern is a prefix of the new pattern (which makes a lot of sense in the context of dynamic user input) etc.

All I'm saying is if we're heading that way, we might also have to replicate these features (if it's even possible with SIMD) to achieve similar performance.

**Reply by saghen** [reply]
Around one and a half years have passed and quite a lot has changed! The library should be ~4x faster than Nucleo (~10x when parallel), it supports unicode, and some convenience APIs like `haystacks.iter().fuzzy_match(needle, config)` have been added. And like before, it supports typos, which [`skim`](https://github.com/skim-rs/skim) has experimented with.

> Incremental inspection of the matcher state while it's running

I thought about adding this but you may find it to be unnecessary when the implementation is 10x faster in parallel. On my (admittedly beefy) machine, with 8 threads, it matches 1.4 million file paths in 3.5ms (see the [Chromium benchmark](https://github.com/saghen/frizbee/blob/main/BENCHMARKS.md#chromium)).

> optimizations when its current pattern is a prefix of the new pattern

This also ended up being largely unnecessary due to the speed-up, with only selectivity of <10% showing gains (https://github.com/saghen/frizbee/pull/65#issuecomment-4724308446), although this code use some more work.

One thing that's still missing is [prefix, substring, suffix, and exact matching support](https://github.com/saghen/frizbee/issues/18) which I can add in 0.11. I'm nearing the point of marking a 1.0 release but would like to see a bit more real world use beforehand. I'd be happy to take a look at creating a PR if this still interests you!

**Reply by alexpasmantier** [reply]
That's terrific news!
Congrats on the speedups, the benchmarks do look very promising 🥳 
Looking forward to experiment with it.
A good way to stress test the swap is to run the text channel on the linux codebase (given the machine has enough ram since that eats up ~11GB) and verifying that tv remains usable.
I'm sure we'll have to refactor a few things since the initial implementation does quite overfit on nucleo but it should be manageable :)

> One thing that's still missing is prefix, substring, suffix, and exact matching support

I do tend to use that very often to filter out noise and I believe it's an important feature to have when you're trying to fuzzy match against a very big corpus with lots of similar haystacks (especially if you're being typo resistant and end up with these additional matches). It's probably a hard requirement for the tv swap but isn't blocking per se to start experimenting!

**Comment by alexpasmantier** [comment]
https://github.com/alexpasmantier/television/pull/1107
